sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/UploadBufferPool.java

```diff
@@ -0,0 +1,172 @@
+    and calling write. This means operations like currentBuf.remaining() will not result in race conditions.
+     */
+    public Flux<ByteBuffer> write(ByteBuffer buf) {
+        // Check if there's a buffer holding any data from a previous call to write. If not, get a new one.
+        if (this.currentBuf == null) {
+            this.currentBuf = this.getBuffer();
+        }
+
+        Flux<ByteBuffer> result;
+        // We can fit this whole write in the buffer we currently have.
+        if (this.currentBuf.remaining() >= buf.remaining()) {
+            this.currentBuf.put(buf);
+            if (this.currentBuf.remaining() == 0) {
+                // Reset the position so that we can read the whole thing then return this buffer.
+                this.currentBuf.position(0);
+                result = Flux.just(this.currentBuf);
+                // This will force us to get a new buffer next time we try to write.
+                this.currentBuf = null;
+            } else {
+                /*
+                We are still filling the current buffer, so we have no data to return. We will return the buffer once it
+                is filled
+                 */
+                result = Flux.empty();
+            }
+        } else {
+            // We will overflow the current buffer and require another one.
+            // Adjust the window of buf so that we fill up currentBuf without going out of bounds.
+            int oldLimit = buf.limit();
+            buf.limit(buf.position() + this.currentBuf.remaining());
+            this.currentBuf.put(buf);
+            // Set the old limit so we can read to the end in the next buffer.
+            buf.limit(oldLimit);
+
+            // Reset the position so we can read the buffer.
+            this.currentBuf.position(0);
+            result =  Flux.just(this.currentBuf);
+
+            /*
+            Get a new buffer and fill it with whatever is left from buf. Note that this relies on the assumption that
+            the source Flux has been split up into buffers that are no bigger than chunk size. This assumption
+            means we'll only have to over flow once, and the buffer we overflow into will not be filled. This is the
+            buffer we will write to on the next call to write().
+             */
+            this.currentBuf = this.getBuffer();
+            this.currentBuf.put(buf);
+        }
+        return result;
+    }
+
+    /*
+    Note that the upload method will be calling write sequentially as there is only one worker reading from the source
+    and calling write. Hence there is only one worker calling getBuffer at any time.
+     */
+    private ByteBuffer getBuffer() {
+        ByteBuffer result;
+        // There are no buffers in the queue and we have space to allocate one.
+        if (this.buffers.isEmpty() && this.numBuffs < this.maxBuffs) {
+            result = ByteBuffer.allocate(this.buffSize);
```

Comment:
I don't think so. Since we enter this block only when the queue is empty, this is effectively doing a buffers.add(newBuff); return buffers.pop(). The buffer will get added to buffers when it gets returned to the pool.